kernel/core/maths/matfuncs.hpp

```diff
@@ -3567,6 +3567,174 @@ static Matrix matrixFilter(const MatFuncData& funcData, const MatFuncErrorInfo&
 }
 
 
+/////////////////////////////////////////////////
+/// \brief Function to perform a circular shift
+/// of the elements in a matrix by n rows.
+///
+/// \param funcData const MatFuncData&
+/// \param errorInfo const MatFuncErrorInfo&
+/// \return Matrix
+///
+/////////////////////////////////////////////////
+static Matrix circShiftRow(const MatFuncData& funcData, const MatFuncErrorInfo& errorInfo)
+{
+    // Inputs: funcData.mat1 funcData.nVal
+    std::vector<mu::value_type> matData = funcData.mat1.data();
+
+    // Circular shift downwards (negative values upwards)
+    // TODO: remove the unnecessary int casts
+    int shift = (int)funcData.nVal % (int)matData.size();
+    shift = - shift;
+    shift = (shift + (int)matData.size()) % (int)matData.size();
+
+    std::rotate(matData.begin(), matData.begin() + shift, matData.end());
+
+    Matrix _mResult = Matrix();
+    _mResult.assign(funcData.mat1.rows(), funcData.mat1.cols(), matData);

```

Comment:
That's the second copy. Is that intended? You can get the internal array as a reference. Furthermore, if the passed matrix has been transposed from somewhere else, it is not working as expected. What about following this approach:
```C++
Matrix _mResult(funcData.mat1);
_mResult.extend(_mResult.rows(), _mResult.cols()); // force a transposition baking here

// Get the reference
std::vector<mu::value_type>& vData = _mResult.data();

// Call std::rotate() ...
return _mResult;
```
Then you won't need to call `extend()` twice in the opposite cases. Just transpose the matrix twice (the second one is the cheap call)